File: app/src/main/java/social/entourage/android/user/edit/UserEditPasswordFragment.java
package social.entourage.android.user.edit;


import android.graphics.Typeface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.text.method.PasswordTransformationMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.Toast;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import social.entourage.android.Constants;
import social.entourage.android.EntourageEvents;
import social.entourage.android.R;
import social.entourage.android.api.model.User;
import social.entourage.android.base.EntourageDialogFragment;

public class UserEditPasswordFragment extends EntourageDialogFragment {

    // ----------------------------------
    // CONSTANTS
    // ----------------------------------

    public static final String TAG = "user_edit_password_fragment";
    public static final int MIN_PASSWORD_LENGTH = 6;
    public static final int MAX_PASSWORD_LENGTH = 6;

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @BindView(R.id.user_old_password)
    EditText oldPasswordEditText;

    @BindView(R.id.user_new_password)
    EditText newPasswordEditText;

    @BindView(R.id.user_confirm_password)
    EditText confirmPasswordEditText;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    public UserEditPasswordFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        View toReturn = inflater.inflate(R.layout.fragment_user_edit_password, container, false);
        ButterKnife.bind(this, toReturn);
        EntourageEvents.logEvent(EntourageEvents.EVENT_SCREEN_09_4);
        configureView();

        return toReturn;
    }

    @Override
    public void onResume() {
        super.onResume();

        if (getDialog() != null && getDialog().getWindow() != null) {
            getDialog().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        }
    }

    // ----------------------------------
    // BUTTONS HANDLING
    // ----------------------------------

    @OnClick(R.id.title_close_button)
    protected void onCloseButton() {
        dismiss();
    }

    @OnClick(R.id.user_edit_password_save_button)
    protected void onSaveButton() {
        if (validatePassword()) {
            EntourageEvents.logEvent(EntourageEvents.EVENT_SCREEN_09_4_SUBMIT);
            UserEditFragment userEditFragment = (UserEditFragment) getFragmentManager().findFragmentByTag(UserEditFragment.TAG);
            if (userEditFragment != null) {
                userEditFragment.saveNewPassword(newPasswordEditText.getText().toString().trim());
            }

            dismiss();
        }
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------

    private void configureView() {
        oldPasswordEditText.setTypeface(Typeface.DEFAULT);
        oldPasswordEditText.setTransformationMethod(new PasswordTransformationMethod());

        newPasswordEditText.setTypeface(Typeface.DEFAULT);
        newPasswordEditText.setTransformationMethod(new PasswordTransformationMethod());

        confirmPasswordEditText.setTypeface(Typeface.DEFAULT);
        confirmPasswordEditText.setTransformationMethod(new PasswordTransformationMethod());

        oldPasswordEditText.requestFocus();
    }

    private boolean validatePassword() {
        String oldPassword = oldPasswordEditText.getText().toString().trim();
        String newPassword = newPasswordEditText.getText().toString().trim();
        String confirmPassword = confirmPasswordEditText.getText().toString().trim();
        String userPassword = "";
        UserEditFragment userEditFragment = (UserEditFragment) getFragmentManager().findFragmentByTag(UserEditFragment.TAG);
        if (userEditFragment != null) {
            User user = userEditFragment.getEditedUser();
            userPassword = user.getSmsCode();
        }
        if (!oldPassword.equals(userPassword)) {
            displayToast(R.string.user_edit_password_invalid_current_password);
            return false;
        }
        if (newPassword.length() < MIN_PASSWORD_LENGTH) {
            displayToast(R.string.user_edit_password_new_password_too_short);
            return false;
        }
        if (!newPassword.equals(confirmPassword)) {
            displayToast(R.string.user_edit_password_not_match);
            return false;
        }

        return true;
    }

    private void displayToast(@StringRes int stringId) {
        Toast.makeText(getActivity(), stringId, Toast.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/social/entourage/android/old_v7/user/edit/UserEditPasswordFragment.kt
package social.entourage.android.old_v7.user.edit

import android.graphics.Typeface
import android.os.Bundle
import android.text.method.PasswordTransformationMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import androidx.annotation.StringRes
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.fragment_user_edit_password.*
import kotlinx.android.synthetic.main.layout_view_title.*
import social.entourage.android.tools.log.AnalyticsEvents
import social.entourage.android.R
import social.entourage.android.base.BaseDialogFragment
import social.entourage.android.tools.view.EntSnackbar

class UserEditPasswordFragment  : BaseDialogFragment() {
    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(R.layout.fragment_user_edit_password, container, false)
    }

    override fun onResume() {
        super.onResume()
        dialog?.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        AnalyticsEvents.logEvent(AnalyticsEvents.EVENT_SCREEN_09_4)
        configureView()
    }

    // ----------------------------------
    // BUTTONS HANDLING
    // ----------------------------------
    fun onCloseButton()  = dismiss()

    fun onSaveButton() {
        if (validatePassword()) {
            AnalyticsEvents.logEvent(AnalyticsEvents.EVENT_SCREEN_09_4_SUBMIT)
            val userEditFragment = parentFragmentManager.findFragmentByTag(UserEditFragment.TAG) as UserEditFragment?
            userEditFragment?.saveNewPassword(user_new_password?.text.toString().trim { it <= ' ' })
            dismiss()
        }
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------
    private fun configureView() {
        user_old_password?.typeface = Typeface.DEFAULT
        user_old_password?.transformationMethod = PasswordTransformationMethod()
        user_new_password?.typeface = Typeface.DEFAULT
        user_new_password?.transformationMethod = PasswordTransformationMethod()
        user_confirm_password?.typeface = Typeface.DEFAULT
        user_confirm_password?.transformationMethod = PasswordTransformationMethod()
        user_old_password?.requestFocus()
        title_close_button.setOnClickListener { onCloseButton() }
        user_edit_password_save_button.setOnClickListener { onSaveButton() }
    }

    private fun validatePassword(): Boolean {
        val oldPassword = user_old_password?.text.toString().trim { it <= ' ' }
        val newPassword = user_new_password?.text.toString().trim { it <= ' ' }
        val confirmPassword = user_confirm_password?.text.toString().trim { it <= ' ' }
        val userEditFragment = parentFragmentManager.findFragmentByTag(UserEditFragment.TAG) as UserEditFragment?
        val userPassword = userEditFragment?.presenter?.editedUser?.smsCode ?: ""
        return when {
            oldPassword != userPassword -> {
                displaySnackbar(R.string.user_edit_password_invalid_current_password)
                false
            }
            newPassword.length < MIN_PASSWORD_LENGTH -> {
                displaySnackbar(R.string.user_edit_password_new_password_too_short)
                false
            }
            newPassword != confirmPassword -> {
                displaySnackbar(R.string.user_edit_password_not_match)
                false
            }
            else -> true
        }
    }

    private fun displaySnackbar(@StringRes stringId: Int) {
        view?.let { EntSnackbar.make(it, stringId, Snackbar.LENGTH_SHORT).show() }
    }

    companion object {
        // ----------------------------------
        // CONSTANTS
        // ----------------------------------
        const val TAG = "user_edit_password_fragment"
        const val MIN_PASSWORD_LENGTH = 6
    }
}